src/library_gl.js

```diff
@@ -665,46 +665,116 @@ var LibraryGL = {
 
-      var prevBlend = gl.getParameter(gl.BLEND);
-      if (prevBlend) gl.disable(gl.BLEND);
+        var prevCullFace = gl.getParameter(gl.CULL_FACE);
+        if (prevCullFace) gl.disable(gl.CULL_FACE);
 
-      var prevCullFace = gl.getParameter(gl.CULL_FACE);
-      if (prevCullFace) gl.disable(gl.CULL_FACE);
+        var prevDepthTest = gl.getParameter(gl.DEPTH_TEST);
+        if (prevDepthTest) gl.disable(gl.DEPTH_TEST);
 
-      var prevDepthTest = gl.getParameter(gl.DEPTH_TEST);
-      if (prevDepthTest) gl.disable(gl.DEPTH_TEST);
+        var prevStencilTest = gl.getParameter(gl.STENCIL_TEST);
+        if (prevStencilTest) gl.disable(gl.STENCIL_TEST);
 
-      gl.drawArrays(gl.TRIANGLE_STRIP, 0, 4);
+        function draw() {
+          gl.vertexAttribPointer(context.blitPosLoc, 2, gl.FLOAT, false, 0, 0);
+          gl.drawArrays(gl.TRIANGLE_STRIP, 0, 4);
+        }
+
+        if (gl.createVertexArray) {
+          // WebGL 2 or OES_vertex_array_object
+          var prevVAO = gl.getParameter(gl.VERTEX_ARRAY_BINDING);
+          var vao = gl.createVertexArray();
+          gl.bindVertexArray(vao);
+          gl.enableVertexAttribArray(context.blitPosLoc);
 
-      if (prevDepthTest) gl.enable(gl.DEPTH_TEST);
-      if (prevCullFace) gl.enable(gl.CULL_FACE);
-      if (prevBlend) gl.enable(gl.BLEND);
+          draw();
 
-      gl.bindTexture(gl.TEXTURE_2D, prevTextureBinding);
-      gl.activeTexture(prevActiveTexture);
-      // prevEnableVertexAttribArray?
-      // prevVertexAttribPointer?
-      gl.bindBuffer(gl.ARRAY_BUFFER, prevVB);
-      gl.useProgram(prevProgram);
+          gl.bindVertexArray(prevVAO);
+          gl.deleteVertexArray(vao);
+        } else {
+          var prevVertexAttribPointer = {
+            buffer: gl.getVertexAttrib(context.blitPosLoc, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING),
+            size: gl.getVertexAttrib(context.blitPosLoc, gl.VERTEX_ATTRIB_ARRAY_SIZE),
+            stride: gl.getVertexAttrib(context.blitPosLoc, gl.VERTEX_ATTRIB_ARRAY_STRIDE),
+            type: gl.getVertexAttrib(context.blitPosLoc, gl.VERTEX_ATTRIB_ARRAY_TYPE),
+            normalized: gl.getVertexAttrib(context.blitPosLoc, gl.VERTEX_ATTRIB_ARRAY_NORMALIZED),
+            pointer: gl.getVertexAttribOffset(context.blitPosLoc, gl.VERTEX_ATTRIB_ARRAY_POINTER),
+          };
+          var maxVertexAttribs = gl.getParameter(gl.MAX_VERTEX_ATTRIBS);
+          var prevVertexAttribEnables = [];
+          for (var i = 0; i < maxVertexAttribs; ++i) {
+            var prevEnabled = gl.getVertexAttrib(i, gl.VERTEX_ATTRIB_ARRAY_ENABLED);
+            var wantEnabled = i == context.blitPosLoc;
+            if (prevEnabled && !wantEnabled) {
+              gl.disableVertexAttribArray(i);
+            }
+            if (!prevEnabled && wantEnabled) {
+              gl.enableVertexAttribArray(i);
+            }
+            prevVertexAttribEnables[i] = prevEnabled;
+          }
+
+          draw();
```

Comment:
We should not need to save and restore full array of vertex attribute enabled state for this operation, just preserve and restore the one vertex attribute index that is used for `blitPosLoc`? Even if the vertex attribute array has any other enabled pointers from any other VBO in it, that should not disturb anything for the blitting, because those would be ignored for the purpose of the shader program that is used for the blit. To my understanding that is safe GL behavior - I have not encountered a scenario where excess enabled attributes would be specced to produce an error.

Reply:
Sorry I forgot to include the reason this was needed. I had to go dig around in my old emails to remember. The problem is this:

WebGL: INVALID_OPERATION: drawArrays: no buffer is bound to enabled attribute

That is, UE is leaving an attribute enabled but without a bound buffer - and that [isn't valid](https://www.khronos.org/registry/webgl/specs/latest/1.0/#6.2).

Reply:
It is not a bug; it's valid for the application to leave GL in that state as long as it doesn't try to draw before fixing it. This state restoration code needs to be able to handle it.

Reply:
Got it, thanks @kainino0x

Reply:
> That is, UE is leaving an attribute enabled but without a bound buffer - and that [isn't valid](https://www.khronos.org/registry/webgl/specs/latest/1.0/#6.2).

This sounds odd, can we come up with an example sequence what UE does that goes against the use here?

Calling e.g.

```c++
glBindBuffer(GL_ARRAY_BUFFER, myVbo);
glVertexAttribPointer(/*index=*/0, size, type, normalized, stride, /*offsetToVbo=*/0);
glVertexAttribPointer(/*index=*/1, size, type, normalized, stride, /*offsetToVbo=*/16);
glEnableVertexAttribArray(0);
glEnableVertexAttribArray(1);
glBindBuffer(GL_ARRAY_BUFFER, 0);
glDrawArrays(...);
```

is valid and correct, even if bound shader program only would use attribute at index 0, but not the attribute at index 1. That is, having extraneous attributes bound is not a bug, and should not affect performance.

When we are doing the blit, we only need one vertex attribute `context.blitPosLoc` for that, so we should be able to just `gl.vertexAttribPointer(context.blitPosLoc)` + `gl.enableVertexAttribArray(context.blitPosLoc)` that one attribute, and then restore the activation state later? If there are any other attributes enabled or disabled for the blit call, those will be ignored because the shader program does not use any of them - and after popping the state, things will be back to normal for all attributes, since we revert the only attribute at `context.blitPosLoc` that we changed?

My understanding is that the section https://www.khronos.org/registry/webgl/specs/latest/1.0/#6.2 says that rendering from client side memory (CPU memory without a VBO) is not allowed, but UE is not doing that, and neither is the blit code?

I.e. section 6.2 says that the following code would be wrong:

```c++
glBindBuffer(GL_ARRAY_BUFFER, 0);
glVertexAttribPointer(/*index=*/0, size, type, normalized, stride, /*offsetToVbo=*/0);
glEnableVertexAttribArray(0);
glDrawArrays(...);
```

but this usage should not happen here?

Reply:
You're right about what 6.2 is talking about. Let me look into this further.

Reply:
The repro case is very simple:

```c
// [don't VertexAttribPointer for attrib 1]
glEnableVertexAttribArray(1);
emscripten_webgl_commit_frame();
```

I've included this as an emscripten test.

Reply:
Essentially, if you do

```js
gl.bindBuffer(gl.ARRAY_BUFFER, buf);
gl.vertexAttribPointer(0, ...);
gl.enableVertexAttribArray(0);
gl.enableVertexAttribArray(1);
gl.drawArrays(...);
```

it is invalid.

Reply:
Yeah, that would cause such a bug - but how does that occur with the blitting code here? The blit needs to enable one attribute index, which it fills with a pointer to a VBO right after, so the blit framebuffer code does not enable any attribute indices without a VBO. UE4 engine does so neither, but it also always uses VBOs. I am not sure how it could happen here that the engine would get in a state where it enabled a vertex attribute array index without having a VBO bound to that index?

Reply:
I don't know enough about UE's GL calls to say why it's leaving an extra attribute enabled. The problem only happens during the "fly through" part of Epic Zen Garden (causing the screen to be black), but once you switch to the "interactive" part, it renders fine.